cctp_primitives/src/proving_system/init.rs

```diff
@@ -18,104 +13,204 @@ use std::sync::{RwLock, RwLockReadGuard};
 /// The parameter `max_degree` is required in order to derive a unique hash for the key itself.
-pub fn load_g2_committer_key(
-    max_degree: usize,
-    supported_degree: usize,
-) -> Result<(), SerializationError> {
-    match load_generators::<G2>(max_degree, supported_degree) {
+pub fn load_g2_universal_params(max_degree: usize) -> Result<(), SerializationError> {
+    {
+        let pp_g2_guard = G2_UNIVERSAL_PARAMS.read().map_err(|_| {
+            SerializationError::IoError(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Failed to acquire lock for G2_UNIVERSAL_PARAMS".to_owned(),
+            ))
+        })?;
+        if pp_g2_guard.is_some() {
+            return Err(SerializationError::IoError(std::io::Error::new(
+                std::io::ErrorKind::AlreadyExists,
+                "G2_UNIVERSAL_PARAMS has already been generated",
+            )));
+        }
+    }
+    match load_universal_params::<G2>(max_degree) {
         // Generation/Loading successful, assign the key to the lazy_static
-        Ok(loaded_key) => {
-            G2_COMMITTER_KEY
+        Ok(loaded_params) => {
+            G2_UNIVERSAL_PARAMS
                 .write()
                 .as_mut()
                 .map_err(|_| {
                     SerializationError::IoError(std::io::Error::new(
                         std::io::ErrorKind::Other,
-                        "G2_COMMITTER_KEY write failed",
+                        "G2_UNIVERSAL_PARAMS write failed",
                     ))
                 })?
-                .replace(loaded_key);
+                .replace(loaded_params);
             Ok(())
         }
         // Error while generating/reading file/writing file
         Err(e) => Err(e),
     }
 }
 
-/// Return a RwLockGuard containing the G1CommitterKey, if G1CommitterKey has been initialized,
-/// otherwise return Error.
-pub fn get_g1_committer_key<'a>(
-) -> Result<RwLockReadGuard<'a, Option<CommitterKeyG1>>, ProvingSystemError> {
-    let ck_g1_guard = G1_COMMITTER_KEY.read().map_err(|_| {
-        ProvingSystemError::Other("Failed to acquire lock for G1 Committer Key".to_owned())
+/// If `G1_UNIVERSAL_PARAMETERS` has been initialized, return `CommitterKeyG1`, otherwise return
+/// Error.
+/// If `supported_degree.is_some()`, then `CommitterKeyG1` is trimmed to the specified size.
+pub fn get_g1_committer_key(
+    supported_degree: Option<usize>,
+) -> Result<CommitterKeyG1, ProvingSystemError> {
+    let pp_g1_guard = G1_UNIVERSAL_PARAMS.read().map_err(|_| {
+        ProvingSystemError::Other("Failed to acquire lock for G1_UNIVERSAL_PARAMS".to_owned())
     })?;
-    if ck_g1_guard.is_some() {
-        Ok(ck_g1_guard)
+
+    if pp_g1_guard.is_some() {
+        let supported_degree =
+            supported_degree.unwrap_or_else(|| pp_g1_guard.as_ref().unwrap().max_degree());
+        let (ck, _) =
+            InnerProductArgPC::<_, Digest>::trim(&pp_g1_guard.as_ref().unwrap(), supported_degree)
```

Comment:
Note for the future: this approach is fine but everytime we trim we do a copy of the generators. Currently the generators, in the CommiterKey struct is a Vec, maybe we can do better by defining them as a slice with some lifetime. For sure this is not a change to be performed here.